### ISSUE TITLE
Update ffmpeg.imageio for compatibility with ffmpeg 3.1

### ISSUE THREAD

**Issue by skycaptain** [OpenImageIO/oiio#1465]
### Problem

In ffmpeg 3.1 the API for getting the AVCodecContext [changed](https://www.ffmpeg.org/doxygen/3.1/structAVStream.html#a12826d21779289356722971d362c583c). Also avcodec_decode_video2 was [deprecated](https://www.ffmpeg.org/doxygen/3.1/deprecated.html#_deprecated000021). I'm about to make a pull request soon.

### Versions

* OIIO branch/version: 1.6.19
* OS: OSX 10.11.5

**Comment by lgritz** [comment]
Yes, I just noticed that recently. I spent a short time trying to fix it, and realized it would be more extensive than I was prepared to do (or knew how to do, my ffmpeg knowledge is close to zero). So I settled for just turning off the deprecation errors by adding  `"-Wno-deprecated-declarations"` compile properties in src/libOpenImageIO/CMakeLists.txt and src/ffmpeg.imageio/CMakeLists.txt.

So I am very happy that you are going to fix it for real.

I did leave behind this:

```
// In ffmpeg 3.1.1, the AVStream->codec field was marked as deprecated,
//we're supposed to use ->codecpar instead.
#define USE_CODECPAR (LIBAVCODEC_VERSION_INT >= AV_VERSION_INT(57,24,0))
```

in ffmpeginput.cpp, please feel free to use that or something like it, to conditionally compile so that the code will build cleanly with both new and old versions of ffmpeg.

Also, with your fix, you should eliminate  the `"-Wno-deprecated-declarations"` compile properties where I added them previously.

**Comment by skycaptain** [comment]
I'm already working on this issue; I think most of the work is done. Just one thing, afaik the codecpar was already deprecated in 3.1:
```c++
// In ffmpeg 3.1, the AVStream->codec field was marked as deprecated,
// we're supposed to use ->codecpar instead.
#define USE_CODECPAR (LIBAVCODEC_VERSION_INT >= AV_VERSION_INT(57, 48, 101))
```
I've altered ffmpeginput.cpp to work with the current code without changing too much. But, as far as I understand, they changed or will change the underlaying loop for reading single frames. We may rewrite the whole thing to be compatible with the new API for a later release (maybe 1.8).

**Comment by lgritz** [comment]
We can't deprecate ffmpeg < 3.1 until we are sure that all the important OIIO users are on 3.1 or later. (In fact, I think my employers are one of those places!) If there are people stuck with earlier releases, we will need to retain the #ifdefs or whatever to make sure it builds cleanly either way.

**Comment by skycaptain** [comment]
OK. As said, we may want to drop the support for far future releases of oiio. But as far the code works with the most common versions of ffmpeg, I'm fine.

**Comment by lgritz** [comment]
Far future, yes. Near future, we may need to keep it working across that API change. If small, that can be managed with a few #if's. If it's a more extensive change, we may wish to clone the ffmpeginput.cpp file so there is one version for the new API and one for the old one (the latter of which can eventually be removed) and logic in the CMakeLists.txt to choose one or the other. See src/jpeg2000.imageio/CMakeLists.txt as an example of how I did this recently for an OpenJPEG version change that was a new API too much trouble to try to disentangle in one file.

**Comment by skycaptain** [comment]
OK. I'll keep that in mind.

**Comment by devernay** [comment]
I think the right check is on LIBAVFORMAT_VERSION_INT, not LIBAVCODEC_VERSION_INT, see https://ffmpeg.org/pipermail/ffmpeg-cvslog/2016-April/099192.html

**Comment by lgritz** [comment]
@skycaptain, how is this work going?

**Comment by skycaptain** [comment]
Ähm, huh. How do I put this, without looking unprofessional. I've made the necessary changes weeks ago; but simply forgot to commit 😞 . Here you go: #1503.

**Comment by lgritz** [comment]
No prob, always glad to get it!

**Comment by skycaptain** [comment]
@devernay: I will check this.

**Comment by lgritz** [comment]
Fixed by #1515
